# Dense Neural Network Classifier for Iris Flower Dataset

## (using Keras with TensorFlow backend)

### Data loading, and pre-procesing
1 - Import data
2 - One-hot encode the flower classes
3 - Create training and testing sets

In [17]:
import pandas as pd
import tensorflow as tf
from keras.models import Sequential
from keras.layers import Dense
from sklearn.model_selection import train_test_split

In [3]:
features = ['sepal-lenght','sepal-width','petal-length','petal-width','type']
rough = pd.read_csv("data/iris.data",names=features )

In [4]:
print(rough.head())

   sepal-lenght  sepal-width  petal-length  petal-width         type
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa


In [5]:
# make a copy of the data

data = rough.copy(deep=True)

In [6]:
# Create one-hot encoding for the flower types

one_hot = pd.get_dummies(data['type'])

In [7]:
# Bring datasets together

data.drop('type', axis=1)
data = data.join(one_hot)
data = data.drop('type', axis=1)

In [8]:
data

sepal-lenght  sepal-width  petal-length  petal-width  Iris-setosa  \
0             5.1          3.5           1.4          0.2            1   
1             4.9          3.0           1.4          0.2            1   
2             4.7          3.2           1.3          0.2            1   
3             4.6          3.1           1.5          0.2            1   
4             5.0          3.6           1.4          0.2            1   
5             5.4          3.9           1.7          0.4            1   
6             4.6          3.4           1.4          0.3            1   
7             5.0          3.4           1.5          0.2            1   
8             4.4          2.9           1.4          0.2            1   
9             4.9          3.1           1.5          0.1            1   
10            5.4          3.7           1.5          0.2            1   
11            4.8          3.4           1.6          0.2            1   
12            4.8          3.0           1.4          0.1            1   
13            4.3          3.0           1.1          0.1            1   
14            5.8          4.0           1.2          0.2            1   
15            5.7          4.4           1.5          0.4            1   
16            5.4          3.9           1.3          0.4            1   
17            5.1          3.5           1.4          0.3            1   
18            5.7          3.8           1.7          0.3            1   
19            5.1          3.8           1.5          0.3            1   
20            5.4          3.4           1.7          0.2            1   
21            5.1          3.7           1.5          0.4            1   
22            4.6          3.6           1.0          0.2            1   
23            5.1          3.3           1.7          0.5            1   
24            4.8          3.4           1.9          0.2            1   
25            5.0          3.0           1.6          0.2            1   
26            5.0          3.4           1.6          0.4            1   
27            5.2          3.5           1.5          0.2            1   
28            5.2          3.4           1.4          0.2            1   
29            4.7          3.2           1.6          0.2            1   
..            ...          ...           ...          ...          ...   
120           6.9          3.2           5.7          2.3            0   
121           5.6          2.8           4.9          2.0            0   
122           7.7          2.8           6.7          2.0            0   
123           6.3          2.7           4.9          1.8            0   
124           6.7          3.3           5.7          2.1            0   
125           7.2          3.2           6.0          1.8            0   
126           6.2          2.8           4.8          1.8            0   
127           6.1          3.0           4.9          1.8            0   
128           6.4          2.8           5.6          2.1            0   
129           7.2          3.0           5.8          1.6            0   
130           7.4          2.8           6.1          1.9            0   
131           7.9          3.8           6.4          2.0            0   
132           6.4          2.8           5.6          2.2            0   
133           6.3          2.8           5.1          1.5            0   
134           6.1          2.6           5.6          1.4            0   
135           7.7          3.0           6.1          2.3            0   
136           6.3          3.4           5.6          2.4            0   
137           6.4          3.1           5.5          1.8            0   
138           6.0          3.0           4.8          1.8            0   
139           6.9          3.1           5.4          2.1            0   
140           6.7          3.1           5.6          2.4            0   
141           6.9          3.1           5.1          2.3            0   
142      

In [9]:
nn_ready = data.values
nn_ready

array([[ 5.1,  3.5,  1.4, ...,  1. ,  0. ,  0. ],
       [ 4.9,  3. ,  1.4, ...,  1. ,  0. ,  0. ],
       [ 4.7,  3.2,  1.3, ...,  1. ,  0. ,  0. ],
       ..., 
       [ 6.5,  3. ,  5.2, ...,  0. ,  0. ,  1. ],
       [ 6.2,  3.4,  5.4, ...,  0. ,  0. ,  1. ],
       [ 5.9,  3. ,  5.1, ...,  0. ,  0. ,  1. ]])


### Split the pre-processed data into training and testing sets
- Use shuffle and split into 80% training and 20% testing


In [10]:
X = data[features[0:4]].values

In [11]:
Y = data[list(data.columns.values)[4:7]].values

In [13]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2, random_state=7)

## Create a dense neural net

### Define the model structure
Five fully connected layers:
1. 4 nodes (input)
2. 10 nodes
3. 15 nodes
4. 5 nodes
5. 3 nodes (output)

Last layer activated with a *softmax* since it is a multiclass classification problem

In [24]:
model = Sequential()
model.add(Dense(10, input_dim=4, activation='relu'))
model.add(Dense(15, activation='relu'))
model.add(Dense(5, activation='relu'))
model.add(Dense(3, activation='softmax'))

Ideally, would use Sigmoidal Cross-Entropy as the loss function for multiclass classification problems

In [25]:
model.compile(loss="categorical_crossentropy", optimizer='adam', metrics=['accuracy'])

### Fit/train the model
Using batch size of 10 entries and 200 epochs

In [26]:
model.fit(x_train, y_train, validation_data=(x_test, y_test), epochs=200, batch_size=10)

Train on 120 samples, validate on 30 samples
Epoch 1/200
120/120 [==============================] - 1s 6ms/step - loss: 1.9517 - acc: 0.3167 - val_loss: 1.4594 - val_acc: 0.4000
Epoch 2/200
120/120 [==============================] - 0s 458us/step - loss: 1.3914 - acc: 0.3167 - val_loss: 1.1424 - val_acc: 0.4000
Epoch 3/200
120/120 [==============================] - 0s 466us/step - loss: 1.1484 - acc: 0.3083 - val_loss: 1.1002 - val_acc: 0.3667
Epoch 4/200
120/120 [==============================] - 0s 463us/step - loss: 1.1208 - acc: 0.3250 - val_loss: 1.1187 - val_acc: 0.3333
Epoch 5/200
120/120 [==============================] - 0s 726us/step - loss: 1.1159 - acc: 0.3250 - val_loss: 1.1075 - val_acc: 0.4333
Epoch 6/200
120/120 [==============================] - 0s 607us/step - loss: 1.1002 - acc: 0.3750 - val_loss: 1.1033 - val_acc: 0.3000
Epoch 7/200
120/120 [==============================] - 0s 367us/step - loss: 1.0979 - acc: 0.2667 - val_loss: 1.1030 - val_acc: 0.2667
Epoch 8/200


Epoch 61/200
120/120 [==============================] - 0s 374us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1097 - val_acc: 0.2333
Epoch 62/200
120/120 [==============================] - 0s 600us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1096 - val_acc: 0.2333
Epoch 63/200
120/120 [==============================] - 0s 407us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1095 - val_acc: 0.2333
Epoch 64/200
120/120 [==============================] - 0s 428us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1097 - val_acc: 0.2333
Epoch 65/200
120/120 [==============================] - 0s 378us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1097 - val_acc: 0.2333
Epoch 66/200
120/120 [==============================] - 0s 331us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1096 - val_acc: 0.2333
Epoch 67/200
120/120 [==============================] - 0s 697us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1095 - val_acc: 0.2333
Epoch 68/200
120/120 [==========================

120/120 [==============================] - 0s 339us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1111 - val_acc: 0.2333
Epoch 122/200
120/120 [==============================] - 0s 399us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1112 - val_acc: 0.2333
Epoch 123/200
120/120 [==============================] - 0s 421us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1112 - val_acc: 0.2333
Epoch 124/200
120/120 [==============================] - 0s 396us/step - loss: 1.0975 - acc: 0.3583 - val_loss: 1.1107 - val_acc: 0.2333
Epoch 125/200
120/120 [==============================] - 0s 412us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1108 - val_acc: 0.2333
Epoch 126/200
120/120 [==============================] - 0s 322us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1111 - val_acc: 0.2333
Epoch 127/200
120/120 [==============================] - 0s 422us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1113 - val_acc: 0.2333
Epoch 128/200
120/120 [==============================] 

120/120 [==============================] - 0s 489us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1114 - val_acc: 0.2333
Epoch 181/200
120/120 [==============================] - 0s 505us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1117 - val_acc: 0.2333
Epoch 182/200
120/120 [==============================] - 0s 417us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1117 - val_acc: 0.2333
Epoch 183/200
120/120 [==============================] - 0s 412us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1115 - val_acc: 0.2333
Epoch 184/200
120/120 [==============================] - 0s 332us/step - loss: 1.0972 - acc: 0.3583 - val_loss: 1.1116 - val_acc: 0.2333
Epoch 185/200
120/120 [==============================] - 0s 377us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1118 - val_acc: 0.2333
Epoch 186/200
120/120 [==============================] - 0s 838us/step - loss: 1.0973 - acc: 0.3583 - val_loss: 1.1116 - val_acc: 0.2333
Epoch 187/200
120/120 [==============================] 

## Summary
Accurary stopped at ~36% and loss stabilized around 1.097. Points to several issues, some of them being very small dataset for a model this size